src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,117 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int ELEMENT_NOT_FOUND = -1;
+    private static final double CAPACITY_MULTIPLIER = 1.5;
+
+    private int size;
+    private T[] array;
+
+    @SuppressWarnings("unchecked")
+    public ArrayList() {
+        this.array = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        resize();
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexForAddMethod(index);
+        resize();
+        System.arraycopy(array, index, array, index + 1, size - index);
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            this.add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
+        checkIndex(index);
 
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedElement = (T) array[index];
+        System.arraycopy(array, index + 1, array, index, size - index - 1);
+        size--;
+
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int elementIndex = findElementIndex(element, array);
+        if (elementIndex == -1) {
+            throw new NoSuchElementException("Invalid element");
+        }
+
+        return remove(elementIndex);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void resize() {
+        if (size + 1 >= array.length) {
```

Comment:
```suggestion
        if (size > array.length) {
```

Reply:
If I rewrite it that way it fails the tests